### PR TITLE
feat: Fetch similar events on basis of location and topic.

### LINKED ISSUE #207: Filter events based on location in similar events section
To be done after #198 gets merged
Working on it :+1:

### PR DESCRIPTION
**Pull request by angmas1**
Fixes #207  

Changes: Similar events are fetched on basis of event topic(as was before) as well as event location.

Screenshots for the change:

### CODE REVIEW

**Comment by addiegupta** [comment]
But Issue #267 is closed?

**Comment by angmas1** [comment]
> But #267 is a PR and it is closed?

Sorry changed

**Comment by angmas1** [comment]
@liveHarshit pls review

**Comment by angmas1** [comment]
@iamareebjamal please review

**Comment by iamareebjamal** [comment]
Ask for peer review first

**Comment by angmas1** [comment]
@liveHarshit @nikit19 please review

**Comment by angmas1** [comment]
@iamareebjamal please review before it gets stale

**Comment by iamareebjamal** [comment]
@liveHarshit @theLimitBreaker @addiegupta Please review. Peer reviews are an important part of OSS

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 1]
```diff
@@ -286,9 +288,13 @@ class EventDetailsFragment : Fragment() {
         rootView.refundPolicy.text = event.refundPolicy
 
         // Similar Events Section
-        if (event.eventTopic != null) {
+        if (event.eventTopic != null || !event.locationName.isNullOrBlank() ||
+            !event.searchableLocationName.isNullOrBlank()) {
             similarEventsContainer.visibility = View.VISIBLE
```
Comment (theLimitBreaker) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: Not related to your PR, but we could use `view.isVisible=true` for following kotlin coding style. Concise and cleaner meaning

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt [hunk 2]
```diff
@@ -17,32 +17,54 @@ class SimilarEventsViewModel(private val eventService: EventService) : ViewModel
...
             .subscribeOn(Schedulers.io())
             .observeOn(AndroidSchedulers.mainThread())
-            .doOnSubscribe({
+            .doOnSubscribe {
                 mutableProgress.value = true
-            }).subscribe({
+            }.subscribe({
                 mutableProgress.value = false
-                mutableSimilarEvents.value = it.filter { it.id != eventId }
+                mutableSimilarIdEvents.value = it.filter { it.id != eventId }
```
Comment (theLimitBreaker) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: That's unnecessary change.

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt [hunk 3]
```diff
@@ -153,4 +168,31 @@ class SimilarEventsFragment : Fragment() {
         moreLikeThis.isGone = similarEvents.isEmpty()
         similarEventsRecycler.isGone = similarEvents.isEmpty()
     }
+
+    private fun setUpAdapter() {
+        similarEvents.clear()
+        var id: Long
+        var add: Boolean
+        if (similarIdEvents.size != 0 && similarLocationEvents.size != 0) {
```
Comment (theLimitBreaker) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: Use `when` expression when there  is more than `if else` expression

Comment (addiegupta) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: I don't think `when` can be used on 2 variables. Please correct me if I'm wrong

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt [hunk 4]
```diff
@@ -153,4 +168,31 @@ class SimilarEventsFragment : Fragment() {
...
+            similarIdEvents.forEach {
+                id = it.id
+                add = true
+                similarLocationEvents.forEach {
+                    if (id == it.id) add = false
```
Comment (addiegupta) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: You can add a `break` statement to prevent iterating after `add` has already been set to `false`

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt [hunk 5]
```diff
@@ -17,32 +17,54 @@ class SimilarEventsViewModel(private val eventService: EventService) : ViewModel
...
+    val similarLocationEvents: LiveData<List<Event>> = mutableSimilarLocationEvents
+    private val mutableSimilarIdEvents = MutableLiveData<List<Event>>()
+    val similarIdEvents: LiveData<List<Event>> = mutableSimilarIdEvents
     private val mutableError = SingleLiveEvent<String>()
     val error: LiveData<String> = mutableError
 
     var eventId: Long = -1
 
-    fun loadSimilarEvents(id: Long) {
+    fun loadIdEvents(id: Long) {
```
Comment (addiegupta) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: Maybe refactor this to `loadSimilarIdEvents` as all other variables also consist of `similarIdEvents` such as the mutable variable

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt (hunk 4 above)
Comment (theLimitBreaker) [inline app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: You could add `break` when `id` is matched 
or
 you would like to use kotlin stdlib `find`  for better readability
```
val notfound = list.find{ id == it.id}==null
if(notFound) addtolist 
```

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt (hunk 3 above)
Comment (theLimitBreaker) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: yes it could . for example
```
when{
   variable1!=0 && variable!=0 ->{ do somwthing here}
    variable3==2 ->{ do something here}
}
```

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt (hunk 1 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: good that's for pointing that out will change

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt (hunk 4 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: yes, thanks didn't notice that.

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt (hunk 5 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: okay, makes sense.

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt (hunk 4 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: resolved

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt (hunk 5 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsViewModel.kt]: resolved

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt (hunk 4 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: resolved

File: app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt (hunk 3 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/topic/SimilarEventsFragment.kt]: resolved

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt (hunk 1 above)
Comment (angmas1) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: resolved

**Comment by aman210697** [comment]
When you update your branch, do rebase instead of merge.
